kip-0019.md

```diff
@@ -0,0 +1,93 @@
+---
+KIP: 0019
+Title: Coin-v6 and fungible-v3
+Author: Robert Soeldner (robert@kadena.io)
+Status: Draft
+Type: Standard
+Category:
+Created: 2023-05-02
+---
+
+## Abstract
+
+The next iteration of the coin contract proposes three main changes to enhance
+the _maintainability_ and _security_ of the contract.
+
+Firstly, the `rotate` function, which is responsible rotating an account guard in a contract that implements the `fungible-v2` interface,
+will be removed. In the case of fungible tokens that also implement the `fungible-xchain` interface, this will result "correct by construction" cross-chain transactions by avoiding the risk of locking funds sent to the receiving account.
```

Comment:
you don't need to rotate to get funds stuck in cross chain. Only specifying the wrong receiver keyset. And you can specify any arbitrary keyset (correct me if wrong), the sending chain can't check if the account exists on the receiving chain and with which keyset it was created. getting stuck doing rotate during crosschain is a special case, i'm not sure that happens in the real world..

Reply:
rotate saved me a few times from such errors. if I specified the wrong keyset during cross chain but still a keyset you have control of. You would fail on the second part but could fix it by rotating the receiver to the matching key when finding out the mistake. Then attempt the continuation part again.

Reply:
the only way that rotate can contribute to eliminating the risk on stuck cross-chain is if you are going to disable cross-chain on non-principal -classic arbitrary- account names and validate the keyset for the principal account name in the first initial step of the defpact for cross-chain. Sorry if that is already implied in the KIP here, I might have missed it then..

Reply:
> you don't need to rotate to get funds stuck in cross chain. Only specifying the wrong receiver keyset is enough. You can specify any arbitrary keyset (correct me if wrong), the sending chain can't check if the account exists on the receiving chain and with which keyset it was created. getting stuck doing rotate during crosschain is a special case, i'm not sure that happens in the real world..

> rotate saved me a few times from such errors. if I specified the wrong keyset during cross chain but still a keyset I have control of. It would fail on the second part but could fix it by rotating the receiver to the matching key when finding out the mistake. Then attempt the continuation part again.

> the only way that removing rotate can contribute to eliminating the risk on stuck cross-chain is if you are going to disable cross-chain on non-principal -classic arbitrary- account names and validate-principal for the principal account name in the first initial step of the defpact for cross-chain. Sorry if that is already implied in the KIP here, I might have missed it then..

Just to be clear here, we're not trying to make the blockchain your mother. If you type in the wrong receiver or their governance details, you will lose your funds unless you have a means of recovery available, like you own the keys and you simply sent to the wrong one of your many accounts. The problems you describe are endemic of any call to a transfer in which you fat finger details. We cannot and do not want to guard you against unforced errors like what you describe - you must take reasonable care. However, we *can* protect against nondeterministic problems such as the person you're sending to being able to rotate in between your crosschain transfers. 

And yes, there is another KIP out to enforce that `create-account` from the next `coin` version onward only works for principal account creation.

Reply:
@trendzetter I'm very happy you brought up this point and started interacting with us. As Emily said, we currently do not protect against typing a wrong receiver, but this KIP aims to increase the security. The additional model annotations will provide more secure properties and trust into the contract.

(I'm on my way creating the new KIP for the `create-account` restriction.)

Reply:
To add color to @rsoeldner - no blockchain *does* protect you against these things. There's always the possibility that a) you fail to transfer because you've used `coin.transfer`, or b) you fat fingered a `coin.transfer-create`, which is not considerd a "safe" transfer for this exact reason. We offer a remedy for this: the `/local` endpoint, which can simulate transactions before you execute the on-chain transaction. Wallets are always encouraged to use this to confirm details for both the data query side (receiver exists, guards match etc.), and the simulated transaction execution side in which the transaction is run against the most recent blockchain state. Automating this workflow is standard. I can only see people running into the problem of fat-fingering if they either execute their own queries and fail to check their details beforehand, or if they're using transfer scripts that fail to do so.